shopelectro/context.py

```diff
@@ -0,0 +1,375 @@
+class AbstractContext(ObjectsComposition, ABC):
+
+    super: 'AbstractContext' = None
+
+    def __init__(  # Ignore PyDocStyleBear
+        self,
+        url_kwargs: typing.Dict[str, str]=None,
+        request: http.HttpRequest=None
+    ):
+        """
+        :param url_kwargs: Came from `urls` module.
+        :param request: Came from `urls` module
+        """
+
+        self.url_kwargs_ = url_kwargs or {}
+        self.request_ = request
+
+    @property
+    def url_kwargs(self) -> typing.Dict[str, str]:
+        return self.url_kwargs_ or self.super.url_kwargs
+
+    @property
+    def request(self) -> http.HttpRequest:
+        return self.request_ or self.super.request
+
+    @abstractmethod
+    def get_context_data(self) -> typing.Dict[str, typing.Any]:
+        ...
+
+
+class AbstractPageContext(AbstractContext, ABC):
+
+    super: 'AbstractPageContext' = None
+
+    def __init__(  # Ignore PyDocStyleBear
+        self,
+        url_kwargs: typing.Dict[str, str]=None,
+        request: http.HttpRequest=None
+    ):
+        """
+        :param url_kwargs: Came from `urls` module.
+        :param request: Came from `urls` module
+        """
+
+        if url_kwargs:
+            assert 'slug' in url_kwargs
+        super().__init__(url_kwargs, request)
+
+    @property
+    def slug(self) -> str:
+        return self.url_kwargs['slug']
+
+    @property
+    @lru_cache(maxsize=1)
+    def page(self):
+        return ModelPage.objects.get(slug=self.slug)
+
+
+class AbstractProductsListContext(AbstractPageContext, ABC):
+
+    super: 'AbstractProductsListContext' = None
+
+    @property
+    def products(self) -> QuerySet:
+        if self.super:
+            return self.super.products
+        else:
+            raise NotImplementedError
+
+
+class Category(AbstractProductsListContext):
+    @property
+    def products(self) -> QuerySet:
+        return models.Product.actives.get_category_descendants(
+            self.page.model
+        )
+
+    def get_context_data(self):
+        """Add sorting options and view_types in context."""
+        # @todo #550:15m Take `view_type` value from dataclass.
+        #  Depends on updating to python3.7
+        view_type = self.request.session.get('view_type', 'tile')
+
+        # @todo #550:60m Create `Product.get_brand`
+        #  Or `ProductQuerySet.get_brand`.
+        group_tags_pairs = (
+            models.Tag.objects
+            .filter_by_products(self.products)
+            .get_group_tags_pairs()
+        )
+
+        return {
+            'products_data': prepare_tile_products(self.products),
+            'group_tags_pairs': group_tags_pairs,
+            # can be `tile` or `list`. Defines products list layout.
+            'view_type': view_type,
+        }
+
+
+class TaggedCategory(AbstractProductsListContext):
+
+    def get_sorting_index(self):
+        return int(self.url_kwargs.get('sorting', 0))
+
+    # @todo #550:30m Test if context.TaggedCategory.get_tags cached.
```

Comment:
What exactly  should be cached here?

Reply:
@artemiy312 , y, i supposed them. But django ORM caches them by default.
So, i rm both lru_cache and task